# Question 3:

## Part (a)

In [2]:
def f(x):
	return((x-1)**6)

def fexpanded(x):
	return(x**6-6*x**5+15*x**4-20*x**3+15*x**2-6*x+1)


def fnested(x):
	result = 1
	for i in range(6):
		result *= (x-1)
	return result

## Part (b)

In [3]:
for i in range(1,8):
	x = 1+10**(-i)
	print(x, f(x), fexpanded(x), fnested(x))
	

1.1 1.0000000000000052e-06 9.999999948107074e-07 1.0000000000000055e-06
1.01 1.0000000000000052e-12 1.000088900582341e-12 1.0000000000000054e-12
1.001 9.999999999993393e-19 -2.6645352591003757e-15 9.999999999993393e-19
1.0001 9.999999999993393e-25 -3.552713678800501e-15 9.99999999999339e-25
1.00001 1.0000000000393072e-30 4.440892098500626e-15 1.0000000000393072e-30
1.000001 9.999999995064001e-37 3.552713678800501e-15 9.999999995064e-37
1.0000001 1.0000000035032031e-42 1.7763568394002505e-15 1.0000000035032031e-42


## Part (c)
We are iterating from 1 to 7 inclusive in the above loop. With each iteration **x** is approaching a number closer to 1 with more precicion needed to specify it as different than 1. The **f(x)** function uses the original function form to evaluate the value of the specified function when given each iterative **x** value. The **fexpanded(x)** uses the expanded definition of the function to evaluate the value at each **x**. The **fnested(x)** uses the nested multiplication form of **f(x)** to evaluate at each **x** value. With each iteration, we can see how each form of the polynomial function evaluates differently from eachother as **x** approaches a number closer to one.\

Using the definition of **f(x)** we can find the exact answers to be:\
0.000001\
1x10<sup>-12</sup>\
1x10<sup>-18</sup>\
1x10<sup>-24</sup>\
1x10<sup>-30</sup>\
1x10<sup>-36</sup>\
1x10<sup>-42</sup>\
I found these numbers by plugging the original **f(x)** definition into Desmos and evaluating at **x** when **x** is equal to each of the specified numbers in the left collumn for the above results.

## Part (d)
The **Definition** and **Nested Multiplication** versions of the function appear to have the most consistent results and agree with eachother's results. The limitations of floating point arrithmatic cause issues for the **Expanded Form** as there is more room for floating point errors when subtraction is involved with arithmetic involving small numbers.

# Question 4:

In [4]:
def fexpandederror(x):
	return(x**6-6*x**5+15*x**4-19.9*x**3+15*x**2-6*x+1)

## Part (a)

In [22]:
x1 = 1.1
print("fexpanded of 1.1 is", fexpanded(x1))
print("fexpandederror of 1.1 is:", fexpandederror(x1))

fexpanded of 1.1 is 9.999999948107074e-07
fexpandederror of 1.1 is: 0.13310099999999725


f(1.1) = 0.000001

## Part (b)

In [21]:
x2 = 1.1
fexpanded_1p1 = fexpanded(x2)
dfexpanded_1p1 = 6*(x2)**5-6*5*(x2)**4+15*4*(1.1)**3-20*3*(x2)**2+15*2*(x2)-6
print("Calculation for derivitve of fexpanded:", dfexpanded_1p1)

conditionNum = abs(((x2)/fexpanded_1p1)*(dfexpanded_1p1))
print("Condition Number is:", conditionNum)

Calculation for derivitve of fexpanded: 6.000000001193939e-05
Condition Number is: 66.00000035562664


## Part (c)
The **Condition Number** of 66.00000035562664 indicates that the function is sensitive to small changes in the function's input.

# Question 5:

In [27]:
sumf=0.0
for i in range(1,10001):
	sumf = sumf + 1.0/i**2
print("sum forward", sumf)

sum forward 1.6448340718480652


In [26]:
sumr=0.0
for i in range(10000, 0, -1):
	sumr = sumr + 1.0 / i**2
print("Sum reverse:", sumr)

Sum reverse: 1.6448340718480596


In [28]:
exact = 1.6448340718480597698
forwardError = abs(sumf - exact)
reverseError = abs(sumr - exact)
print("Forward Error:", forwardError)
print("Reverse Error:", reverseError)

Forward Error: 5.329070518200751e-15
Reverse Error: 2.220446049250313e-16


The **Forward Sum** has a lower error than the **Reverse Sum**. This is likely due to less rounding errors in the **Forward Sum** because it adds the smaller numbers with more precision before adding the larger numbers. When going in reverse order, the iteratively smaller added sums get rounded off.

# Question 6:

In [29]:
def g(x):
	return(4*x*(1-x))
p = 1.0/7.0
for i in range(1,41):
	p = g(p)
print("first p is", p)

p2 = 1.0/7.0 + 10**(-12)
for i in range(1,41):
	p2 = g(p2)
print("second p is", p2)

first p is 0.08068495147915851
second p is 0.9190923575057521


## Part (a)
The code is iteratively performing an equation involving multiplication and subtraction on two different, but very close numbers. The two answers should be very close.

## Part (b)
The error is coming from iterative rounding causing the numbers to compound the errors with each iteration. The arithmetic is unstable from the rounding causing chaotic behavior from just a very small change of input sizes.